Added LeafShears tool and small fix

Aleksey-Terzi opened the pull request:
1) Added LeafShears which allows to clear leaves fast and without loosing durability on this special shears when breaking leaves. Default 'clear cube' has size 3 x 3 x 3. Could be configured.
New tool is requiring Citadel plugin.

2) Fixed error in LoreTag.appliedOn method when item's lore = null

Aleksey-Terzi committed:
Added LeafShears tool and fix of error in LoreTag.appliedOn when item's lore = null

ProgrammerDan commented on src/main/java/com/github/maxopoly/WurstCivTools/effect/LeafShears.java:
```diff
@@ -0,0 +1,118 @@
+		
+		Bukkit.getScheduler().runTask(WurstCivTools.getPlugin(), new Runnable() {
+            public void run() {
+            	handItem.setDurability((short)(handItem.getDurability() - 1));
```
Wrap the durability prevention in a config option; we'll probably aim for a tool that does AOE for only one durability cost vs an infinite tool. It will be repairable so even if expensive not a dubious burden to maintain.

Aleksey-Terzi replied:
This part of code is not preventing decreasing durability.  
Following lines:

handItem.setDurability((short)(handItem.getDurability() + 1));  

...  

handItem.setDurability((short)(handItem.getDurability() - 1));  

done just to synchronize client with server.  
When event is cancelled - this is automatically preventing from changing durability.  

But unfortunately there is bug in Bukkit and client is thinking that durability was changed (but it wasn't) and therefore after some time - client will "tell" player that tool was broken, but if player will refresh inventory - tool will be restored.  

This is very old Bukkit's bug: https://bukkit.atlassian.net/browse/BUKKIT-1825

ProgrammerDan replied:
Yes I'm aware of that. Please reread my comment :P

ProgrammerDan replied:
Sorry, misstated in my comment, meant to say -- "Yes I'm aware this only resyncs the client -- add it in a config option because I want to force durability loss even though event is cancelled".

Aleksey-Terzi replied:
Ok...but this is additional problems - if tool has Unbreakable enchantment - then I will need to introduce little more code to correctly process this.

Also if you want to force durability loss - we need to decided how tool will loose this durability.
Right now - if I'm using shears  - they will be broken almost immediately after I start clearing leaves.  

This will not allow to clear big amount of leaves we have in Eilon.

Remember experience with custom leaves in Devoted - internal regions of southern continent were almost unpopulated, because of people haven't possibility to clear forest (after fire spreading was disabled), but iron on Devoted was very and very cheap.

Please let me know how you think this should work.

ProgrammerDan replied:
Instead of cancelling the event, can we just clear the drops and force air as type? Then AOE would proceed w/o durability loss and only the block hit would cause a durability loss. Unbreaking would be applicable then, with no extra code.

Aleksey-Terzi replied:
Yes, I think this is doable and this should much simplify code.

ProgrammerDan replied:
Whew, ok. Definitely worth trying. So then the stuff you have now -- wrap as a config "prevent_durability_loss" or something -- and the path suggested as the alternate if durability loss is allowed. 

We currently anticipate allowing durability loss offset by the lossless AOE as a good "middle ground" giving good clearing ability but not infinite uses :).

Aleksey-Terzi replied:
Ok, seems this is really not simple task to clear "block drop" without cancellation of event.

Therefore I suggest to use of following ways:

1) Ignore unbreaking enhcnatment when breaking leaves

2) Simulate Unbreaking by custom calculation of durability loss, i.e. use this table: http://minecraft.gamepedia.com/Enchanting#Unbreaking

I would choose second variant.

Please let me know your opinion.

ProgrammerDan replied:
Ah, shame. I suppose 2, although it's just another soft-point for maintenance on version upgrades. It'll do!

Aleksey-Terzi replied:
Ok, will prepare new version during today.

> although it's just another soft-point for maintenance on version upgrades.

Unfortunately yes.

Aleksey-Terzi replied:
I think we could have here one more problem - plugin should control if tool broken or no.

Therefore I have one more question.  

Bukkit has some method: net.minecraft.server.v1_10_R1.ItemStack.damage

This method is doing damage to the tool correctly with statistics, checking enchantments, checking unbreakable tag, calling events, tool breaking  and etc, the only two issues:

1) it is version-dependent

2) variable "handle" is private in CraftItemStack, so I will need to use reflection to get it (additional compatibility issue)

So, the question.
Will you accept code if I will use this method? (I could just "hardcode" version for now and later will make modules for Bukkit versioning like I did in Orebfuscator)

Or do you prefer to be compatible with namespaces as much as possible and I should avoid using NMS?

EDIT: so I'm not sure if Bukkit will break tool himself or no when durability will reach limit (I think no, but ...), I could test this later today.

ProgrammerDan replied:
It's likely that we'd need to add versioned compatibility late on, so go for it.

Aleksey-Terzi replied:
Cool, thank you.

ProgrammerDan commented on src/main/java/com/github/maxopoly/WurstCivTools/effect/LeafShears.java:
```diff
@@ -0,0 +1,118 @@
+		}
+		
+		e.setCancelled(true);
+		e.getBlock().setType(Material.AIR);
```
Pretty sure this needs to be wrapped in the Citadel check; abandon the whole attempt if this block is citadel protected on an inaccessible group.

Maxopoly replied:
Also we dont want to cancel the event if the player doesnt have bypass, so it still applies normal reinforcement damage

ProgrammerDan replied:
Yeah, that would suffice here, good catch. For the AOE a bit more is required to trigger citadel breaks on all the connected blocks, I left a comment there but if you remember the exact method in Citadel that would be a help I'm sure.

Aleksey-Terzi replied:
> Pretty sure this needs to be wrapped in the Citadel check; abandon the whole attempt if this block is citadel protected on an inaccessible group.  
> 
> Also we dont want to cancel the event if the player doesnt have bypass, so it still applies normal reinforcement damage

This is already working correctly.  
When Citadel is preventing block breaking because of another group or /ctb disabled - it is cancelling event.

But since WurstCivTools's BlockBreak event has HIGH priority and since WurstCivTools has softdepend on Citadel - then if Citadel is cancelling this event - it is even not firing in WurstCivTools.  

You can test this.

ProgrammerDan replied:
I'd still recommend adding the check. 

Event priorities change; they should not be depended on with us crazy Civ programmers at the helm :)

Aleksey-Terzi replied:
I don't think this is good idea, because of code will duplicate work of Citadel's break event (remember about lags)  

Also Bastion plugin is depending on priorities as well (method BastionListener.onBlockBreak).  Therefore if change in priorities will be done in Citadel - this will break Bastion and maybe other plugins as well.  

Also currently Citadel haven't public method to process this and therefore I will have two choices: either fix Citadel or duplicate Citadel's code in WurstCivTools which is not good for compatibility.

ProgrammerDan replied:
Convinced. But, make an issue on Citadel's code to put a comment there to _never_ change that priority :D

Aleksey-Terzi replied:
Ok, will do!

ProgrammerDan commented on src/main/java/com/github/maxopoly/WurstCivTools/effect/LeafShears.java:
```diff
@@ -0,0 +1,118 @@
+		PlayerReinforcement playerRein = (PlayerReinforcement)rein;
+		
+		return playerRein.canBypass(player)
+			|| player.hasPermission("citadel.admin.bypassmode");
```
The way this is constructed, it doesn't remove the reinforcement if bypass is on, nor does it fire the break event; this code as it is will create ghost reinforcements (not a good thing!) so ideally should do one of two things:

1) remove the bypass code; if reinforced just doesn't work, perhaps a message telling players to pick a diff tool.

2) fire the correct method to trigger a full bypass (including reinf material refund) within Citadel.

Aleksey-Terzi replied:
Good catch, I will fix this.

Btw I've found one more bug while tested it is not correctly removing leaves in radius (instead of startZ - used startY) - I will fix this as well.

Aleksey-Terzi replied:
> 1) remove the bypass code; if reinforced just doesn't work, perhaps a message telling players to pick a diff tool.  

I think this is the most ideal variant - because of simple in implementation and we are providing this tool to clear large amount of custom leaves but not bypass reinforcement,

Would be this ok if plugin will tell to player following: "Some of leaves were reinforced and therefore not broken. Please use another tool to break them."

?

ProgrammerDan replied:
Yes! Ideally make it configurable line, and wrap it in the Color.alternateColorCode call (forgetting exact imprint) so that we can configure in a color as well.

Aleksey-Terzi committed:
Bug fixes and added chance of durability loss

Aleksey-Terzi commented:
Fixed bugs and implemented following features:

1) Introduced new config parameter "cannot_bypass_message" - specifies message that player will see when leaf(s) in the cube is not possible to remove because of it is reinforced

2) New config parameter "durability_loss_chance" - is specifying chance to loss durability.
0 - durability never will be loss
1- durability always will be loss

This parameter is working separately from Unbreaking enchantment, therefore resulting chance to loss durability will be = durability_loss_chance \* "Unbreaking chance"

Aleksey-Terzi committed:
Small fix

Maxopoly commented:
Looks awesome, sorry for adding conflicts to the PR, but I wanted to get in the other first, because it had been floating around for too long already. Once those conflicts are fixed I'd suggest we merge this, unless Dan has any further objections.

Aleksey-Terzi committed:
Merged

Aleksey-Terzi committed:
Merge branch 'Civcraft-master'

Aleksey-Terzi commented:
Done, conflicts resolved.

ProgrammerDan commented:
I have no objections, lets get it on testing and confirm live behavior.
